app/src/main/java/ch/hikemate/app/model/authentication/FirebaseAuthRepository.kt

```diff
@@ -127,4 +134,91 @@ class FirebaseAuthRepository : AuthRepository {
     Firebase.auth.signOut()
     onSuccess()
   }
+
+  override fun deleteAccount(
+      password: String,
+      activity: Activity,
+      onSuccess: () -> Unit,
+      onErrorAction: (Exception) -> Unit
+  ) {
+    val user = FirebaseAuth.getInstance().currentUser
+    if (user != null) {
+      val email = user.email ?: throw Exception("User email is null")
+      Log.d("DeleteAccount", "User email: '$email', password: '$password'")
+
+      // Re-authenticate the user
+      reauthenticate(
+          user,
+          password,
+          activity,
+          {
+            Log.d("DeleteAccount", "User re-authenticated")
+
+            // Delete the user profile and saved hikes
+            FirebaseFirestore.getInstance()
+                .collection(ProfileRepositoryFirestore.PROFILES_COLLECTION)
+                .document(user.uid)
+                .delete()
+            FirebaseFirestore.getInstance()
+                .collection(SavedHikesRepositoryFirestore.SAVED_HIKES_COLLECTION)
+                .document(user.uid)
+                .delete()
+
+            Log.d("DeleteAccount", "Profile deleted")
+
+            // Delete the user account
+            user.delete().addOnCompleteListener { task ->
+              if (task.isSuccessful) {
+                Log.d("DeleteAccount", "User account deleted")
+                onSuccess()
+              } else {
+                onErrorAction(task.exception ?: Exception())
+              }
+            }
+          },
+          { onErrorAction(Exception("Error re-authenticating user")) })
+    } else {
+      onErrorAction(Exception("No user is currently signed in"))
+    }
+  }
+
+  override fun isEmailProvider(user: FirebaseUser): Boolean {
+    val providers = user.providerData.map { it.providerId }
+    return providers.contains(EmailAuthProvider.PROVIDER_ID)
```

Comment:
Nitpicking: Don't you have a more performance efficient way to do it? Currently, this implementation is θ(n^2) in CPU usage (not accounting for comparison). I'm confident you can check for this in a θ(n) way

Reply:
Actually it's θ(n) not θ(n^2) but I will change it for θ(n)